<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to segtest Version 1.0 • segtest</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to segtest Version 1.0">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">segtest</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/intro_1.html">Introduction to segtest Version 1.0</a></li>
    <li><a class="dropdown-item" href="../articles/intro_2.html">Introduction to segtest Version 2.0</a></li>
    <li><a class="dropdown-item" href="../articles/parallel.html">Parallel Testing</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/dcgerard/segtest/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to segtest Version 1.0</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/segtest/blob/HEAD/vignettes/intro_1.Rmd" class="external-link"><code>vignettes/intro_1.Rmd</code></a></small>
      <div class="d-none name"><code>intro_1.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dcgerard.github.io/segtest/">segtest</a></span><span class="op">)</span></span></code></pre></div>
<p>This is the introduction to Version 1.0 of segtest. This version was
only applicable to tetraploids. Version 2.0 is applicable to any even
ploidy level.</p>
<p>The <code>segtest</code> package contains a suite of functions to
test and evaluate segregation distortion in F1 populations of
tetraploids. We allow for various types of polyploids (auto, allo, and
segmental) without having the user specify the type of polyploid they
are studying. We also account for genotype uncertainty through the use
of genotype likelihoods, which can be obtained through many genotyping
programs (like <code>updog</code>, <code>fitpoly</code>, and
<code>polyRAD</code>). Details of these methods may be found in Gerard
et al. (2025a). Methods for higher ploidies may be found in Gerard et
al. (2025b). The main functions are:</p>
<ul>
<li>
<code><a href="../reference/multi_lrt.html">multi_lrt()</a></code>: Run any of the likelihood ratio tests for
segregation distortion in parallel across many SNPs.</li>
<li>
<code>multidog_to_g</code>: Format the genotyping output from
<code><a href="https://rdrr.io/pkg/updog/man/multidog.html" class="external-link">updog::multidog()</a></code> to be compatible withe input of
<code><a href="../reference/multi_lrt.html">multi_lrt()</a></code>.</li>
<li>
<code><a href="../reference/lrt_men_g4.html">lrt_men_g4()</a></code>: Likelihood ratio test for segregation
distortion using known genotypes.</li>
<li>
<code><a href="../reference/lrt_men_gl4.html">lrt_men_gl4()</a></code>: Likelihood ratio test for segregation
distortion using genotype likelihoods.</li>
<li>
<code><a href="../reference/offspring_gf_2.html">offspring_gf_2()</a></code>: Offspring genotype frequencies under
the two parameter model of meiosis.</li>
<li>
<code><a href="../reference/offspring_gf_3.html">offspring_gf_3()</a></code>: Offspring genotype frequencies under
the three parameter model of meiosis.</li>
<li>
<code><a href="../reference/simf1g.html">simf1g()</a></code>: Simulate genotype counts from an F1
population of tetraploids.</li>
<li>
<code><a href="../reference/simf1gl.html">simf1gl()</a></code>: Simulate genotype likelihoods from an F1
population of tetraploids.</li>
</ul>
<p>Here, we will demonstrate some of our functions.</p>
<div class="section level2">
<h2 id="offspring-genotype-frequencies">Offspring genotype frequencies<a class="anchor" aria-label="anchor" href="#offspring-genotype-frequencies"></a>
</h2>
<p>We can obtain offspring genotype frequencies via
<code><a href="../reference/offspring_gf_2.html">offspring_gf_2()</a></code> and <code><a href="../reference/offspring_gf_3.html">offspring_gf_3()</a></code>. These
are two different parameterizations of the same model for meiosis. For
<code><a href="../reference/offspring_gf_3.html">offspring_gf_3()</a></code>, you insert the following parameters:</p>
<ul>
<li>
<code>tau</code>: Probability of quadrivalent formation.</li>
<li>
<code>beta</code>: Probability of double reduction given
quadrivalent formation</li>
<li>
<code>gamma1</code>: Probability of <code>AA_aa</code> pairing in
parent 1 given bivalent formation. Only applicable when
<code>p1 = 2</code>.</li>
<li>
<code>gamma2</code>: Probability of <code>AA_aa</code> pairing in
parent 2 given bivalent formation. Only applicable when
<code>p2 = 2</code>.</li>
<li>
<code>p1</code>: The first parent’s genotype.</li>
<li>
<code>p2</code>: The second parent’s genotype.</li>
</ul>
<p>Let’s generate some example genotype frequencies. You can play around
with the parameter values yourself.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/offspring_gf_3.html">offspring_gf_3</a></span><span class="op">(</span></span>
<span>  tau <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  beta <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">6</span>, </span>
<span>  gamma1 <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">3</span>,</span>
<span>  gamma2 <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, </span>
<span>  p1 <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  p2 <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">4</span>,</span>
<span>  y <span class="op">=</span> <span class="va">gf</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"h"</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"Genotype"</span>, </span>
<span>  ylab <span class="op">=</span> <span class="st">"Frequency"</span>,</span>
<span>  ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_1_files/figure-html/unnamed-chunk-2-1.png" alt="A probability mass function of genotypes of a tetraploid F1 population" width="700"></p>
<p>The <code><a href="../reference/offspring_gf_3.html">offspring_gf_3()</a></code> function is safer to use because
there is a dependence between the preferential pairing parameter and the
double reduction rate that bounds these values in
<code><a href="../reference/offspring_gf_2.html">offspring_gf_2()</a></code>, and so in the two-parameter model you
might accidentally choose values that are impossible. I did not set up
checks for these values because the bounds depend on the maximum rate of
double reduction, which can vary significantly. Please see Gerard et
al. (2025) for details.</p>
</div>
<div class="section level2">
<h2 id="when-the-null-is-true">When the null is true<a class="anchor" aria-label="anchor" href="#when-the-null-is-true"></a>
</h2>
<p>We’ll first simulate some data where the null of no segregation
distortion is true.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">g1</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">g2</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">/</span><span class="fl">6</span></span>
<span><span class="va">xi1</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">/</span><span class="fl">3</span></span>
<span><span class="va">xi2</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">/</span><span class="fl">3</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">20</span></span>
<span><span class="va">rd</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simf1g.html">simf1g</a></span><span class="op">(</span></span>
<span>  n <span class="op">=</span> <span class="va">n</span>, </span>
<span>  g1 <span class="op">=</span> <span class="va">g1</span>, </span>
<span>  g2 <span class="op">=</span> <span class="va">g2</span>, </span>
<span>  alpha <span class="op">=</span> <span class="va">alpha</span>, </span>
<span>  xi1 <span class="op">=</span> <span class="va">xi1</span>, </span>
<span>  xi2 <span class="op">=</span> <span class="va">xi2</span><span class="op">)</span></span>
<span><span class="va">gl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simf1gl.html">simf1gl</a></span><span class="op">(</span></span>
<span>  n <span class="op">=</span> <span class="va">n</span>, </span>
<span>  rd <span class="op">=</span> <span class="va">rd</span>, </span>
<span>  g1 <span class="op">=</span> <span class="va">g1</span>,</span>
<span>  g2 <span class="op">=</span> <span class="va">g2</span>, </span>
<span>  alpha <span class="op">=</span> <span class="va">alpha</span>, </span>
<span>  xi1 <span class="op">=</span> <span class="va">xi1</span>,</span>
<span>  xi2 <span class="op">=</span> <span class="va">xi2</span><span class="op">)</span></span></code></pre></div>
<p>The LRT has a large
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value,
which is appropriate since there is no segregation distortion.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lrt_men_g4.html">lrt_men_g4</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, g1 <span class="op">=</span> <span class="va">g1</span>, g2 <span class="op">=</span> <span class="va">g2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lout</span><span class="op">$</span><span class="va">p_value</span></span>
<span><span class="co">#&gt; [1] 0.5698342</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lout_gl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lrt_men_gl4.html">lrt_men_gl4</a></span><span class="op">(</span>gl <span class="op">=</span> <span class="va">gl</span>, g1 <span class="op">=</span> <span class="va">g1</span>, g2 <span class="op">=</span> <span class="va">g2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lout_gl</span><span class="op">$</span><span class="va">p_value</span></span>
<span><span class="co">#&gt; [1] 0.6369666</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="when-the-alternative-is-true">When the alternative is true<a class="anchor" aria-label="anchor" href="#when-the-alternative-is-true"></a>
</h2>
<p>When we simulate data where the alternative is true, we get a very
small
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Multinom.html" class="external-link">rmultinom</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">20</span>, prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">lout</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lrt_men_g4.html">lrt_men_g4</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, g1 <span class="op">=</span> <span class="va">g1</span>, g2 <span class="op">=</span> <span class="va">g2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lout</span><span class="op">$</span><span class="va">p_value</span></span>
<span><span class="co">#&gt; [1] 3.5306e-07</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div class="section level3">
<h3 id="references-1">References<a class="anchor" aria-label="anchor" href="#references-1"></a>
</h3>
<p>Gerard D, Thakkar M, &amp; Ferrão LFV (2025a). “Tests for segregation
distortion in tetraploid F1 populations.” <em>Theoretical and Applied
Genetics</em>, <em>138</em>(30), p. 1–13. <a href="https://doi.org/10.1007/s00122-025-04816-z" class="external-link">doi:10.1007/s00122-025-04816-z</a>.</p>
<p>Gerard, D, Ambrosano, GB, Pereira, GdS, &amp; Garcia, AAF (2025b).
“Tests for segregation distortion in higher ploidy F1 populations.”
<em>G3: Genes | Genomes | Genetics</em>, p. jkaf212. <a href="https://doi.org/10.1093/g3journal/jkaf212" class="external-link">doi:10.1093/g3journal/jkaf212</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Gerard, Mira Thakkar, NSF DBI 2132247.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
