<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to segtest Version 2.0 • segtest</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to segtest Version 2.0">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">segtest</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/intro_1.html">Introduction to segtest Version 1.0</a></li>
    <li><a class="dropdown-item" href="../articles/intro_2.html">Introduction to segtest Version 2.0</a></li>
    <li><a class="dropdown-item" href="../articles/parallel.html">Parallel Testing</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/dcgerard/segtest/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Introduction to segtest Version 2.0</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/segtest/blob/HEAD/vignettes/intro_2.Rmd" class="external-link"><code>vignettes/intro_2.Rmd</code></a></small>
      <div class="d-none name"><code>intro_2.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dcgerard.github.io/segtest/">segtest</a></span><span class="op">)</span></span></code></pre></div>
<p>The <code>segtest</code> package offers a suite of tools for testing
segregation distortion in F1 polyploid populations across diverse
meiotic models. These methods support autopolyploids (full polysomic
inheritance), allopolyploids (full disomic inheritance), and segmental
allopolyploids (partial preferential pairing). Double reduction is
optionally modeled fully in tetraploids and partially (at simplex loci
only) in higher ploidies. A user-specified maximum proportion of
outliers allows the method to accommodate moderate double reduction at
non-simplex loci. Offspring genotypes may be known or modeled using
genotype likelihoods to account for genotype uncertainty. Parent data
may or may not be provided, at your option. Parents can have different
(even) ploidies, at your option. Details of the methods may be found in
Gerard et al. (2025a) and Gerard et al. (2025b).</p>
<p>Additional functions include those that generate gamete and genotype
frequencies under different models of meiosis, functions that simulate
genotype (log) likelihoods, and “competing” tests for segregation
distortion.</p>
<p>The main functions are:</p>
<ul>
<li>
<code><a href="../reference/seg_multi.html">seg_multi()</a></code>: Run the likelihood ratio test for
segregation distortion in parallel at many loci.</li>
<li>
<code>multidog_to_g</code>: Format the genotyping output from
<code><a href="https://dcgerard.github.io/updog/reference/multidog.html" class="external-link">updog::multidog()</a></code> to be compatible with the input of
<code><a href="../reference/seg_multi.html">seg_multi()</a></code>.</li>
<li>
<code><a href="../reference/seg_lrt.html">seg_lrt()</a></code>: Test for segregation distortion for any even
ploidy.</li>
<li>
<code><a href="../reference/gamfreq.html">gamfreq()</a></code>: Gamete frequencies.</li>
<li>
<code><a href="../reference/gf_freq.html">gf_freq()</a></code>: Genotype frequencies of an F1 population of
polyploids.</li>
<li>
<code><a href="../reference/drbounds.html">drbounds()</a></code>: Upper bounds on the double reduction
rate(s) based on two different extreme models of meiosis.</li>
<li>
<code><a href="../reference/simgl.html">simgl()</a></code>: Simulate genotype log-likelihoods given a
vector of genotype counts.</li>
</ul>
<div class="section level2">
<h2 id="gamete-genotype-frequencies">Gamete genotype frequencies<a class="anchor" aria-label="anchor" href="#gamete-genotype-frequencies"></a>
</h2>
<p><code><a href="../reference/gamfreq.html">gamfreq()</a></code> will generate gamete frequencies under
different models of meiosis. <code><a href="../reference/gf_freq.html">gf_freq()</a></code> will generate
genotype frequencies under the same models by convolving the output from
<code><a href="../reference/gamfreq.html">gamfreq()</a></code>. We focus on <code><a href="../reference/gamfreq.html">gamfreq()</a></code>, as
<code><a href="../reference/gf_freq.html">gf_freq()</a></code> uses the same models but applied separately to
each parent.</p>
<p>For autopolyploids, specify <code>type = "polysomic"</code> and,
optionally, the amount of double reduction via <code>alpha</code>.
<code>alpha</code> is a vector of length <code>floor(ploidy / 4)</code>
where element <code>i</code> is the probability a gamete has
<code>i</code> pairs of identical by double reduction alleles. The upper
bounds for <code>alpha</code> can be found via <code><a href="../reference/drbounds.html">drbounds()</a></code>.
E.g., for a parental octoploid with genotype 4 with no and moderate
levels of double reduction:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/drbounds.html">drbounds</a></span><span class="op">(</span>ploidy <span class="op">=</span> <span class="fl">8</span><span class="op">)</span> <span class="co">## DR bounds</span></span>
<span><span class="co">#&gt; [1] 0.38571429 0.02142857</span></span>
<span><span class="fu"><a href="../reference/gamfreq.html">gamfreq</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">4</span>, ploidy <span class="op">=</span> <span class="fl">8</span>, type <span class="op">=</span> <span class="st">"polysomic"</span><span class="op">)</span> <span class="co">## no DR</span></span>
<span><span class="co">#&gt; [1] 0.01428571 0.22857143 0.51428571 0.22857143 0.01428571</span></span>
<span><span class="fu"><a href="../reference/gamfreq.html">gamfreq</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">4</span>, ploidy <span class="op">=</span> <span class="fl">8</span>, alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.01</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"polysomic"</span><span class="op">)</span> <span class="co">## Some DR</span></span>
<span><span class="co">#&gt; [1] 0.022 0.232 0.492 0.232 0.022</span></span></code></pre></div>
<p>For allopolyploids, the possible gamete frequencies can be found in
<code>seg</code>. E.g., for a parental octoploid with genotype 4, the
possible gamete frequencies are</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seg</span><span class="op">[</span><span class="va">seg</span><span class="op">$</span><span class="va">ploidy</span> <span class="op">==</span> <span class="fl">8</span> <span class="op">&amp;</span> <span class="va">seg</span><span class="op">$</span><span class="va">g</span> <span class="op">==</span> <span class="fl">4</span> <span class="op">&amp;</span> <span class="va">seg</span><span class="op">$</span><span class="va">mode</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"disomic"</span>, <span class="st">"both"</span><span class="op">)</span>, <span class="st">"p"</span><span class="op">]</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] 0.0625 0.2500 0.3750 0.2500 0.0625</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] 0.00 0.25 0.50 0.25 0.00</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; [1] 0 0 1 0 0</span></span></code></pre></div>
<p>Note that you also need to filter for the <code>mode</code> to be
either <code>"disomic"</code> or <code>"both"</code> (both disomic and
polysomic). The total number of possible allopolyploid distributions is
<code><a href="../reference/n_pp_mix.html">n_pp_mix()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/n_pp_mix.html">n_pp_mix</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">4</span>, ploidy <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 3</span></span></code></pre></div>
<p>You can specify one of these distributions via a 1-of-3 vector.
E.g.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gamfreq.html">gamfreq</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">4</span>, ploidy <span class="op">=</span> <span class="fl">8</span>, gamma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.0625 0.2500 0.3750 0.2500 0.0625</span></span>
<span><span class="fu"><a href="../reference/gamfreq.html">gamfreq</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">4</span>, ploidy <span class="op">=</span> <span class="fl">8</span>, gamma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.00 0.25 0.50 0.25 0.00</span></span>
<span><span class="fu"><a href="../reference/gamfreq.html">gamfreq</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">4</span>, ploidy <span class="op">=</span> <span class="fl">8</span>, gamma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0 0 1 0 0</span></span></code></pre></div>
<p>Segmental allopolyploids are mixtures of the possible allopolyploid
segregation distributions. E.g., an equal mixture of the three for an
octoploid with genotype 4 is</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gamfreq.html">gamfreq</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">4</span>, ploidy <span class="op">=</span> <span class="fl">8</span>, gamma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">/</span><span class="fl">3</span>, type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] 0.02083333 0.16666667 0.62500000 0.16666667 0.02083333</span></span></code></pre></div>
<p>At simplex, loci, there is only one possible allopolyploid
segregation distribution:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/n_pp_mix.html">n_pp_mix</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">1</span>, ploidy <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span><span class="fu"><a href="../reference/gamfreq.html">gamfreq</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">1</span>, ploidy <span class="op">=</span> <span class="fl">8</span>, gamma <span class="op">=</span> <span class="fl">1</span>, type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.5 0.5 0.0 0.0 0.0</span></span>
<span><span class="fu"><a href="../reference/n_pp_mix.html">n_pp_mix</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">7</span>, ploidy <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span><span class="fu"><a href="../reference/gamfreq.html">gamfreq</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">7</span>, ploidy <span class="op">=</span> <span class="fl">8</span>, gamma <span class="op">=</span> <span class="fl">1</span>, type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.0 0.0 0.0 0.5 0.5</span></span></code></pre></div>
<p>You can account for double reduction at these loci by including
<code>beta</code>. The upper bound of which can be found via
<code><a href="../reference/beta_bounds.html">beta_bounds()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/beta_bounds.html">beta_bounds</a></span><span class="op">(</span>ploidy <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.05357143</span></span>
<span><span class="fu"><a href="../reference/gamfreq.html">gamfreq</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">1</span>, ploidy <span class="op">=</span> <span class="fl">8</span>, gamma <span class="op">=</span> <span class="fl">1</span>, beta <span class="op">=</span> <span class="fl">0.03</span>, type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.53 0.44 0.03 0.00 0.00</span></span>
<span><span class="fu"><a href="../reference/gamfreq.html">gamfreq</a></span><span class="op">(</span>g <span class="op">=</span> <span class="fl">7</span>, ploidy <span class="op">=</span> <span class="fl">8</span>, gamma <span class="op">=</span> <span class="fl">1</span>, beta <span class="op">=</span> <span class="fl">0.03</span>, type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.00 0.00 0.03 0.44 0.53</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="simulating-data">Simulating data<a class="anchor" aria-label="anchor" href="#simulating-data"></a>
</h2>
<p>Let’s suppose we have some genotype frequencies we want to simulate
individual data from:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gf_freq.html">gf_freq</a></span><span class="op">(</span></span>
<span>  p1_g <span class="op">=</span> <span class="fl">2</span>, </span>
<span>  p1_ploidy <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  p1_gamma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.7</span>, <span class="fl">0.3</span><span class="op">)</span>, </span>
<span>  p1_type <span class="op">=</span> <span class="st">"mix"</span>,</span>
<span>  p2_g <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  p2_ploidy <span class="op">=</span> <span class="fl">6</span>, </span>
<span>  p2_gamma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span>, </span>
<span>  p2_type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">gf</span>, type <span class="op">=</span> <span class="st">"h"</span>, xlab <span class="op">=</span> <span class="st">"Genotype"</span>, ylab <span class="op">=</span> <span class="st">"Frequency"</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_2_files/figure-html/unnamed-chunk-9-1.png" alt="Bar plot of genotype frequencies." width="700"></p>
<p>To simulate genotype counts, just use <code>multinom()</code> from
the <code>stats</code> package. Let’s simulate data from 10
individuals.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Multinom.html" class="external-link">rmultinom</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">10</span>, prob <span class="op">=</span> <span class="va">gf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x</span></span>
<span><span class="co">#&gt; [1] 0 0 2 4 4 0 0</span></span></code></pre></div>
<p>To simulate genotype (log) likelihoods, insert these genotype counts
into <code><a href="../reference/simgl.html">simgl()</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simgl.html">simgl</a></span><span class="op">(</span>nvec <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span><span class="va">gl</span></span>
<span><span class="co">#&gt;             [,1]      [,2]      [,3]      [,4]      [,5]       [,6]       [,7]</span></span>
<span><span class="co">#&gt;  [1,]  -2.476575 -1.178405 -2.395080 -4.380673 -7.348080 -12.283402 -26.730748</span></span>
<span><span class="co">#&gt;  [2,] -10.226929 -2.738521 -1.514758 -1.614191 -2.750519  -5.504672 -16.089299</span></span>
<span><span class="co">#&gt;  [3,] -13.071998 -3.968154 -1.970660 -1.450388 -1.970660  -3.968154 -13.071998</span></span>
<span><span class="co">#&gt;  [4,] -16.089299 -5.504672 -2.750519 -1.614191 -1.514758  -2.738521 -10.226929</span></span>
<span><span class="co">#&gt;  [5,] -19.319654 -7.362628 -3.867459 -2.118237 -1.394945  -1.826320  -7.525849</span></span>
<span><span class="co">#&gt;  [6,] -16.089299 -5.504672 -2.750519 -1.614191 -1.514758  -2.738521 -10.226929</span></span>
<span><span class="co">#&gt;  [7,] -19.319654 -7.362628 -3.867459 -2.118237 -1.394945  -1.826320  -7.525849</span></span>
<span><span class="co">#&gt;  [8,] -16.089299 -5.504672 -2.750519 -1.614191 -1.514758  -2.738521 -10.226929</span></span>
<span><span class="co">#&gt;  [9,] -22.827562 -9.587669 -5.365836 -3.006407 -1.654578  -1.273137  -4.947896</span></span>
<span><span class="co">#&gt; [10,] -13.071998 -3.968154 -1.970660 -1.450388 -1.970660  -3.968154 -13.071998</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="testing-for-segregation-distortion">Testing for segregation distortion<a class="anchor" aria-label="anchor" href="#testing-for-segregation-distortion"></a>
</h2>
<p>You can test for segregation distortion using <code><a href="../reference/seg_lrt.html">seg_lrt()</a></code>.
E.g., let’s test for it using the data (both known genotypes and
genotype likelihoods) we simulated from the previous section:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## With known genotypes</span></span>
<span><span class="va">sout1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seg_lrt.html">seg_lrt</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, p1_ploidy <span class="op">=</span> <span class="fl">6</span>, p2_ploidy <span class="op">=</span> <span class="fl">6</span>, p1 <span class="op">=</span> <span class="fl">2</span>, p2 <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">sout1</span><span class="op">$</span><span class="va">p_value</span></span>
<span><span class="co">#&gt; [1] 0.5820762</span></span>
<span><span class="co">## With genotype likelihoods</span></span>
<span><span class="va">sout2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/seg_lrt.html">seg_lrt</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">gl</span>, p1_ploidy <span class="op">=</span> <span class="fl">6</span>, p2_ploidy <span class="op">=</span> <span class="fl">6</span>, p1 <span class="op">=</span> <span class="fl">2</span>, p2 <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">sout2</span><span class="op">$</span><span class="va">p_value</span></span>
<span><span class="co">#&gt; [1] 0.5860578</span></span></code></pre></div>
<p>My recommendation is to always use the genotype log-likelihoods. But
<code><a href="../reference/seg_lrt.html">seg_lrt()</a></code> allows for known genotypes, if that situation
works best for you.</p>
<p>The default (<code>model = "seg"</code>) is to assume your organism
is a segmental allopolyploid, and to account for possible double
reduction at simplex loci. But you should absolutely use other models if
you have more information on your organism:</p>
<ul>
<li>
<code>"seg"</code>: General segmental allopolyploid with possible
double reduction at simplex loci.</li>
<li>
<code>"allo_pp"</code>: Segmental allopolyploid with complete
bivalent pairing (no double reduction).</li>
<li>
<code>"allo"</code>: Pure allopolyploid (disomic inheritance).</li>
<li>
<code>"auto"</code>: Pure autopolyploid with complete bivalent
pairing (no double reduction).</li>
<li>
<code>"auto_dr"</code>: Pure autopolyploid with possible multivalent
pairing (some double reduction).</li>
<li>
<code>"auto_allo"</code>: Same null hypothesis as in polymapR.</li>
</ul>
<p>We allow for some non-valid genotypes via the <code>ob</code>
argument. This is the upper bound on the proportion of outliers. By
default, this is set to 0.03. You can set this to <code>0</code> (or set
<code>outlier = FALSE</code>) if you want any outliers to indicate
segregation distortion.</p>
<p>Make sure that the log-likelihoods are base
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>e</mi><annotation encoding="application/x-tex">e</annotation></semantics></math>.
If they are base 10, you’ll get the wrong
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>-value:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gl10</span> <span class="op">&lt;-</span> <span class="va">gl</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/seg_lrt.html">seg_lrt</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">gl10</span>, p1_ploidy <span class="op">=</span> <span class="fl">6</span>, p2_ploidy <span class="op">=</span> <span class="fl">6</span>, p1 <span class="op">=</span> <span class="fl">2</span>, p2 <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">$</span><span class="va">p_value</span></span>
<span><span class="co">#&gt; [1] 0.9141411</span></span></code></pre></div>
<p>Don’t mess with the technical arguments (<code>ntry</code>,
<code>opt</code>, <code>optg</code>, <code>df_tol</code>). These have to
do with the optimization and how to approximate the degrees of freedom
of the test. Except possibly <code>ntry</code>. You could increase that
if you are seeing weird results. But then let me know, because I haven’t
seen any bad behavior with <code>ntry = 3</code> (the default).</p>
<div class="section level3">
<h3 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<p>Gerard D, Thakkar M, &amp; Ferrão LFV (2025a). “Tests for segregation
distortion in tetraploid F1 populations.” <em>Theoretical and Applied
Genetics</em>, <em>138</em>(30), p. 1–13. <a href="https://doi.org/10.1007/s00122-025-04816-z" class="external-link">doi:10.1007/s00122-025-04816-z</a>.</p>
<p>Gerard, D, Ambrosano, GB, Pereira, GdS, &amp; Garcia, AAF (2025b).
“Tests for segregation distortion in higher ploidy F1 populations.”
<em>G3: Genes | Genomes | Genetics</em>, p. jkaf212. <a href="https://doi.org/10.1093/g3journal/jkaf212" class="external-link">doi:10.1093/g3journal/jkaf212</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Gerard, Mira Thakkar, NSF DBI 2132247.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
