<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Gamete frequencies under a generalized model — gamfreq • segtest</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Gamete frequencies under a generalized model — gamfreq"><meta name="description" content="Returns the gamete frequencies for autopolyploids, allopolyploids,
and segmental allopolyploids, accounting for the effects of double
reduction and partial preferential pairing."><meta property="og:description" content="Returns the gamete frequencies for autopolyploids, allopolyploids,
and segmental allopolyploids, accounting for the effects of double
reduction and partial preferential pairing."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">segtest</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/intro_1.html">Introduction to segtest Version 1.0</a></li>
    <li><a class="dropdown-item" href="../articles/intro_2.html">Introduction to segtest Version 2.0</a></li>
    <li><a class="dropdown-item" href="../articles/parallel.html">Parallel Testing</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/dcgerard/segtest/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Gamete frequencies under a generalized model</h1>
      <small class="dont-index">Source: <a href="https://github.com/dcgerard/segtest/blob/HEAD/R/higher_freq.R" class="external-link"><code>R/higher_freq.R</code></a></small>
      <div class="d-none name"><code>gamfreq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Returns the gamete frequencies for autopolyploids, allopolyploids,
and segmental allopolyploids, accounting for the effects of double
reduction and partial preferential pairing.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gamfreq</span><span class="op">(</span></span>
<span>  <span class="va">g</span>,</span>
<span>  <span class="va">ploidy</span>,</span>
<span>  gamma <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  alpha <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  beta <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mix"</span>, <span class="st">"polysomic"</span><span class="op">)</span>,</span>
<span>  add_dr <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-g">g<a class="anchor" aria-label="anchor" href="#arg-g"></a></dt>
<dd><p>Parent genotype.</p></dd>


<dt id="arg-ploidy">ploidy<a class="anchor" aria-label="anchor" href="#arg-ploidy"></a></dt>
<dd><p>Parent ploidy. Should be even, and between 2 and 20 (inclusive).
Let me know if you need the ploidy to be higher. I can update
the package really easily.</p></dd>


<dt id="arg-gamma">gamma<a class="anchor" aria-label="anchor" href="#arg-gamma"></a></dt>
<dd><p>The mixture proportions for the pairing configurations.
The proportions are in the same order the configurations in
<code><a href="seg.html">seg</a></code>. See Gerard et al (2018) for details on
pairing configurations.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>The double reduction rate(s) (if using). Defaults to 0's.</p></dd>


<dt id="arg-beta">beta<a class="anchor" aria-label="anchor" href="#arg-beta"></a></dt>
<dd><p>The double reduction adjustment for simplex markers if
<code>type = "mix"</code> and <code>add_dr = TRUE</code>. Assumed to be 0 by default.</p></dd>


<dt id="arg-type">type<a class="anchor" aria-label="anchor" href="#arg-type"></a></dt>
<dd><p>Either <code>"mix"</code>, meaning a mixture model of
pairing configurations, or <code>"polysomic"</code> for polysomic inheritance.</p></dd>


<dt id="arg-add-dr">add_dr<a class="anchor" aria-label="anchor" href="#arg-add-dr"></a></dt>
<dd><p>A logical. If <code>type = "polysomic"</code>, then the double
double reduction rate (<code>"alpha"</code>) will be used no matter the
value of <code>add_dr</code>, so set <code>alpha = 0</code> if you don't want it.
But if <code>type = "mix"</code> then we will incorporate double reduction
only in simplex markers (where it matters the most, and where
preferential pairing does not operate).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The vector of gamete frequencies. Element <code>i</code> is the
probability a gamete has genotype <code>i - 1</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="models">Models<a class="anchor" aria-label="anchor" href="#models"></a></h2>



<p>If <code>type = "polysomic"</code>, then the gamete frequencies correspond
to those of Huang et al (2019). Those formulas are for general multiallelic
loci, so see also Appendix G of Gerard (2022) for special case of
biallelic loci. The relevant parameter is <code>alpha</code>, a vector of
length <code>floor(ploidy / 4)</code>, where <code>alpha[[i]]</code> is the
probability that there are <code>i</code> pairs of double reduced alleles
in a gamete. The theoretical upper bound on <code>alpha</code> is given in
<code><a href="drbounds.html">drbounds</a>()</code>.</p>
<p>If <code>type = "mix"</code> and <code>add_dr = FALSE</code>, then the gamete
frequencies correspond to the pairing configuration model of
Gerard et al (2018). This model states that the gamete frequencies are
a convex combination of the disomic inheritance frequencies. The weights
of this convex combination are provided in the <code>gamma</code> parameter. The
total number of disomic segregation patterns is given by
<code><a href="n_pp_mix.html">n_pp_mix</a>()</code>. The order of these segregation patterns used is
the order in <code><a href="seg.html">seg</a></code>.</p>
<p>The model for <code>type = "mix"</code> and <code>add_dr = TRUE</code> is the same
as for <code>type = "mix"</code> and <code>add_dr = FALSE</code> <em>except</em> at
parental simplex loci. At such loci, there are no effects of preferential
pairing, and so the option <code>add_dr = TRUE</code> allows for the effects
of double reduction at simplex loci. The relevant parameter here is
<code>beta</code>. The first three gamete frequencies at simplex loci are
<code>c(0.5 + beta, 0.5 - 2 * beta, beta)</code>, and the rest are 0. The
upper bound on beta for two different models are given by
<code><a href="beta_bounds.html">beta_bounds</a>()</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>

<ul><li><p>Gerard, D. (2023). Double reduction estimation and equilibrium tests in natural autopolyploid populations. <em>Biometrics</em>, 79(3), 2143-2156.</p></li>
<li><p>Gerard, D., Ferrão, L. F. V., Garcia, A. A. F., &amp; Stephens, M. (2018). Genotyping polyploids from messy sequencing data. <em>Genetics</em>, 210(3), 789-807.</p></li>
<li><p>Huang, K., Wang, T., Dunn, D. W., Zhang, P., Cao, X., Liu, R., &amp; Li, B. (2019). Genotypic frequencies at equilibrium for polysomic inheritance under double-reduction. <em>G3: Genes, Genomes, Genetics</em>, 9(5), 1693-1706.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>David Gerard</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Various duplex models</span></span></span>
<span class="r-in"><span><span class="fu">gamfreq</span><span class="op">(</span>g <span class="op">=</span> <span class="fl">2</span>, ploidy <span class="op">=</span> <span class="fl">4</span>, gamma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0 1 0</span>
<span class="r-in"><span><span class="fu">gamfreq</span><span class="op">(</span>g <span class="op">=</span> <span class="fl">2</span>, ploidy <span class="op">=</span> <span class="fl">4</span>, gamma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.25 0.50 0.25</span>
<span class="r-in"><span><span class="fu">gamfreq</span><span class="op">(</span>g <span class="op">=</span> <span class="fl">2</span>, ploidy <span class="op">=</span> <span class="fl">4</span>, gamma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span>, type <span class="op">=</span> <span class="st">"mix"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.125 0.750 0.125</span>
<span class="r-in"><span><span class="fu">gamfreq</span><span class="op">(</span>g <span class="op">=</span> <span class="fl">2</span>, ploidy <span class="op">=</span> <span class="fl">4</span>, alpha <span class="op">=</span> <span class="fl">0</span>, type <span class="op">=</span> <span class="st">"polysomic"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.1666667 0.6666667 0.1666667</span>
<span class="r-in"><span><span class="fu">gamfreq</span><span class="op">(</span>g <span class="op">=</span> <span class="fl">2</span>, ploidy <span class="op">=</span> <span class="fl">4</span>, alpha <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">6</span>, type <span class="op">=</span> <span class="st">"polysomic"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.2222222 0.5555556 0.2222222</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Various simplex models</span></span></span>
<span class="r-in"><span><span class="fu">gamfreq</span><span class="op">(</span>g <span class="op">=</span> <span class="fl">1</span>, ploidy <span class="op">=</span> <span class="fl">4</span>, beta <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">24</span>, gamma <span class="op">=</span> <span class="fl">1</span>, type <span class="op">=</span> <span class="st">"mix"</span>, add_dr <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.54166667 0.41666667 0.04166667</span>
<span class="r-in"><span><span class="fu">gamfreq</span><span class="op">(</span>g <span class="op">=</span> <span class="fl">1</span>, ploidy <span class="op">=</span> <span class="fl">4</span>, alpha <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">6</span>, type <span class="op">=</span> <span class="st">"polysomic"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.54166667 0.41666667 0.04166667</span>
<span class="r-in"><span><span class="fu">gamfreq</span><span class="op">(</span>g <span class="op">=</span> <span class="fl">1</span>, ploidy <span class="op">=</span> <span class="fl">4</span>, gamma <span class="op">=</span> <span class="fl">1</span>, type <span class="op">=</span> <span class="st">"mix"</span>, add_dr <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.5 0.5 0.0</span>
<span class="r-in"><span><span class="fu">gamfreq</span><span class="op">(</span>g <span class="op">=</span> <span class="fl">1</span>, ploidy <span class="op">=</span> <span class="fl">4</span>, alpha <span class="op">=</span> <span class="fl">0</span>, type <span class="op">=</span> <span class="st">"polysomic"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.5 0.5 0.0</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Gerard, Mira Thakkar, NSF DBI 2132247.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

